Add functions for summarizing the memory layout.

adfernandes opened the pull request:
When working with hex files, it's often useful to summarize whether or not they have an entry point, and what the contiguous segments are.

This pull request adds said functions, somewhat like the output of `srec_info` from the `srecord` project.

(A separate patch cleans up whitespace; my editor did it automatically, feel free to ignore it... I cherry picked the patches out, just in case.)

Thanks for a great project, by the way!

adfernandes commented:
Actually, I may change this from JSON to YAML output (or add it as an option).

The output of tools like `size` and `srec_info` are all plain text, so YAML is probably a better choice.

I'll do this, rebase, and re-submit in a day, if you don't mind...

bialix commented:
Thank you for your patch, Andrew.
I won't accept EOL changes, they break annotation for no other benefits, sorry.

Your patch seems OK, but despite your code I need some more: 
* something to put into README, 
* also will be nice to have some short text with use-case scenario so I can put something to the user manual as well.
* I will be very grateful if you will provide some unit tests, so your code won't be broken accidentally. I can write unittests myself, I just need from you couple of examples of hex files and what your new function should report about them.

bialix commented:
Also, about reports. Maybe it would be better to don't put formatter to base IntelHex class, but instead provide helper function-formatter (one or more), i.e.

def hex_info_as_json(ih, indent=2):
    import json
    return json.dumps(ih.info(), ....)

So it can be used as example for other people who want to implement their own formatter. It's really trivial task, I'd like to not spoil core IntelHex library with such formatters. I also have command-line scripts, maybe formatters would be better to put there.

bialix commented:
OK, I found example of srec_info on the page: http://linux.die.net/man/1/srec_examples
Now I understand what it should print. If I'd made it command-line utility, maybe I'd add option `--info` or `--summary` to script hex2dump. That looks at least very close in term of data.

adfernandes commented:
Okay, thanks for the feedback, Alexander!

Re: whitespace changes - no problem... I think I've figured out how to make git ignore whitespace changes in a submodule.

Re formatters, unit tests, etc... no problem, will do. I'll find time to implement what you suggest in the next couple of days, then I'll update the pull request for your review.

bialix commented:
Thank you, feel free to ask if you need help.

adfernandes commented:
Okay, done; I've added everything you described earlier, including tests and documentation.

I decided to split off the YAML dump to its own utility because it just didn't fit in `hexdump`.

All trailing-whitespace changes were reverted to the original repo.

bialix commented:
Brilliant, thanks! I'll merge it later this week.

adfernandes commented:
Yeah - fixed. (Thanks for catching, I hate fixing typos after commits...) :+1: :8ball: 

I also rebased the commit to make it cleaner.

adfernandes commented:
Hi, @bialix, not to bother you or anything... but is there a problem merging this pull request? If there's anything I can do to help, please let me know!

bialix commented:
Andrew, I'm very sorry for such late reply. The RL went in, then summer... I'm sorry again.
I usually trying to test my codebase several python platforms before mark the work as done. Actually I have almost all possible python installations from 2.4 up to 3.x.
Unfortunately your patch only pass all test on Python 2.7 (32/64 bits). On earlier versions there is no assertIsInstance, on 3.x there is no xrange. With xrange it's easy to use helper method from compat.py, but with assertIsInstance I've never worked before, but I'm sure it's easy to fix as well.
I'd try to address this issues hopefully this weekend to help this code land.
If you have any questions about the problems with tests above - please ask.

Check presence of python 2.6-64bit ... Python 2.6.6
   Running tests against 2.6-64bit ... FAILED
running test
...........................................E......................................................................
======================================================================
ERROR: test_segments (intelhex.test.TestIntelHex)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "E:\work\OpenSource\intelhex.git\intelhex\intelhex\test.py", line 769, in test_segments
    self.assertIsInstance(sg, list)
AttributeError: 'TestIntelHex' object has no attribute 'assertIsInstance'
----------------------------------------------------------------------
Ran 114 tests in 0.232s

FAILED (errors=1)

**************************************************

Check presence of python 3.2-32bit ... Python 3.2.5
   Running tests against 3.2-32bit ... FAILED
running test
...........................................E......................................................................
======================================================================
ERROR: test_segments (intelhex.test.TestIntelHex)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "E:\work\OpenSource\intelhex.git\intelhex\intelhex\test.py", line 772, in test_segments
    sg = ih.segments()
  File "E:\work\OpenSource\intelhex.git\intelhex\intelhex\__init__.py", line 859, in segments
    return([xrange(addresses[0], addresses[0]+1)])
NameError: global name 'xrange' is not defined
----------------------------------------------------------------------
Ran 114 tests in 0.597s

FAILED (errors=1)

bialix commented:
OK, I see https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertIsInstance
it should be easy to change assertIsInstance on assertTrue(isinstance(a,b))

bialix commented:
Hi Andrew, I've easily fixed problem with assertIsInstance, but second problem with xrange is not so easy. Due to problems with xrange itself on python 2.x I have very complex code to support long ints in compat.py, and therefore that is hard to use properly in your tests, esp. the line

        self.assertIsInstance(sg[0], xrange)

cannot be easily transormed.

So I'd like to ask you why do you need xrange object in the first place? Maybe it would be safer to switch to just tuples of (start,end) without applying xrange to them?

adfernandes commented:
Hi, @bialix... it's my turn to take days to reply... :smile:!

Hmm... I've been thinking about this, and unfortunately I don't have enough python expertise to understand the issue, exactly.

I used `xrange` objects because they seemed to be ideal representations of allocated memory, without the storage overhead.

I had no idea that they were buggy or incompatible with other representations.

I'm not second guessing you, but what is the issue with python and long ints? I would have suspected that since ints are variable sized in python that they'd be fully supported in all places that can take an int... but like I said, I'm relatively unsophisticated in my python.

I'm not wedded to `xrange` and using tuples is fine with me; I didn't do it originally because they felt "kludgey". But whatever.

Do you want to make the change, or shall I?

(Seriously - could you explain the issue regarding long ints? I'm thinking rather than a tuple, to make a "compat xrange" object, that preserves the semantics of `xrange` but is compatible with your long-int code and requirements...)

Thanks!

adfernandes commented:
Ah... okay. A bit of googling revealed that python2's `range` was removed, and for python3 `xrange` was renamed to `range`... so **that**'s the issue, correct?

If that's right, then I'm assuming the compatibility problem as a standard pythonic way of being backwards compatible?

bialix commented:
About problem with xrange https://bugs.launchpad.net/intelhex/+bug/1408934

After thinking about this more I'm inclined to fix the problem by changing code to use tuples. They are easy to make xrange from them, and can be reused many times, in contrast to xranges. And this won't require any special efforts for testing.

If you have a minute to change this that'd be great. I'll try to make another round on this patch later this week. I have a branch with assertIsInstance fix, you can merge it to your.

Probably you can also suggest using intelhex.compat.range_g (cross-python-version equivalent of xrange which is safe for big ints) in examples of iterating over segments addresses.

adfernandes commented:
Okay. I think that works. I changed the `xrange` objects to tuples, as per your suggestion. It was definitely the easiest thing to do.

I also changed the `isInstance` stuff, as required.

And I rebased the commit.

Please test (I could only test 2.7 and 3.4), and let me know! :+1:

bialix commented:
All tests pass now, thank you!

adfernandes commented:
Great news! :+1: 

It's funny, I often see the children (err, I mean "students") complain about the work that can go into supporting old code... then I realize it is only because they haven't been alive long enough for software to move through its lifecycle.

Heck, I still remembering if it was worth porting some of my perl3 to perl5... :older_man: :laughing:

c1728p9 commented:
Thanks for the change @adfernandes !  This will definitely be nice to have.  Up till now I have been having to extract the address ranges manually.
